### ISSUE TITLE
Three questions regarding image preprocessing, cross entropy and lambda in loss function

### ISSUE THREAD

**Issue by ethanyanjiali** [zzh8829/yolov3-tf2#137]
Hi zzh, I've got three more questions when I read the code:

1) It seems like the only image preprocessing you did is just to resize. Is it on purpose? Usually it requires tons of data augmentation to fully exploit the dataset. And I'm thing if it's necessary to pad the image to keep the aspect ratio. What's your thought on this?

2) Also, the class loss is calculated through sparse categorical entropy instead of binary cross entropy. do you choose to this multiclass loss because it's more stable and easier to converge?

3) I didn't see the lambda_coord and lambda_noobj in your loss function. How do you managed to overcome the problem that noobj is dominating the loss? Or is the lambda not really useful?
 
Thanks.

**Comment by zzh8829** [comment]
Hi great questions, the code here is pretty much true to paper without much modification for performance. 

1. Naive resize is the minimum required transform to train the network at all. Data augmentation is not implemented here but you could easily plug in your own algo. For sure it's always a good idea to augment your dataset for any image related dataset, but I didn't think that's in the scope of this project. In order to keep aspect ratio, some form of cropping or padding is required and they are not easy to implement correctly. Also resizing without keeping the aspect ratio is technically one form of image augmentation, so it's not completely a terrible idea to use them. For more robust real training, I have used this repo in the past https://github.com/aleju/imgaug

2. When each region only produces single class, categorical entropy and sum of binary entropy are equivalent. So if you datasets includes multiple overlapping bounding box then categorical entropy won't performance really well and you should use sum of binary entropy in that case. I choose entropy because in most case training sets are single class.

3. I did some quick search on google looks like lambda_coord and lambda_noobj were included in the previous version of yolo but not yolov3. To be honest, i understand why they existed before but im not sure why they are not needed anymore. Maybe because yolov3 converges much better so theres no need to fine tune these loss scaling parameters.

**Comment by ethanyanjiali** [comment]
Got it. Thanks for the quick reply!

It seems like the author used `letterbox_image` to keep the aspect ratio when training so I was wondering how much gain would it bring. Also, do you know where does the author mentioned about not using lambda in the paper? In fact I have to retrospect to v1 and v2 paper to get a full picture of v3 loss.

**Comment by asquare92** [comment]
> Hi great questions, the code here is pretty much true to paper without much modification for performance.
> 
> 1. Naive resize is the minimum required transform to train the network at all. Data augmentation is not implemented here but you could easily plug in your own algo. For sure it's always a good idea to augment your dataset for any image related dataset, but I didn't think that's in the scope of this project. In order to keep aspect ratio, some form of cropping or padding is required and they are not easy to implement correctly. Also resizing without keeping the aspect ratio is technically one form of image augmentation, so it's not completely a terrible idea to use them. For more robust real training, I have used this repo in the past https://github.com/aleju/imgaug
> 2. When each region only produces single class, categorical entropy and sum of binary entropy are equivalent. So if you datasets includes multiple overlapping bounding box then categorical entropy won't performance really well and you should use sum of binary entropy in that case. I choose entropy because in most case training sets are single class.
> 3. I did some quick search on google looks like lambda_coord and lambda_noobj were included in the previous version of yolo but not yolov3. To be honest, i understand why they existed before but im not sure why they are not needed anymore. Maybe because yolov3 converges much better so theres no need to fine tune these loss scaling parameters.

I am training my dataset consists of 9 classes and there are almost 5 multiple overlapping bounding boxes in a single image. Whenever I try to use Binary_croossentropy, I dimension mismatch error.  If understood you correctly, Binary_crossentropy should work for more then two classes, right?
Furthermore, I am able to train successfully but my val loss doesn't converge after 18.7 and starts to increase no matter what the learning rate is. I first used transfer darknet then transfer fine_tune, but no improvements. Can Binary_crossentropy lead to improvement if I am able to use it? Any help in this regard is appreciated.
